libpromises/eval_context.c

```diff
@@ -3484,6 +3484,47 @@ bool EvalContextIsIgnoringLocks(const EvalContext *ctx)
     return ctx->ignore_locks;
 }
 
+StringSet *ClassesMatchingLocalRecursive(const EvalContext *ctx, const char* regex, const Rlist *tags, bool first_only, size_t stack_index)
+{
+  assert(ctx != NULL);
+  StackFrame *frame = SeqAt(ctx->stack, stack_index);
+  StringSet *matches;
+  if (frame->type == STACK_FRAME_TYPE_BUNDLE)
+  {
+    ClassTableIterator *iter = ClassTableIteratorNew(frame->data.bundle.classes, frame->data.bundle.owner->ns, false, true); // from EvalContextClassTableIteratorNewLocal()
+    matches = ClassesMatching(ctx, iter, regex, tags, first_only);
+    ClassTableIteratorDestroy(iter);
+  }
+  else
+  {
+    matches = StringSetNew(); // empty for passing up the recursion chain
+  }
+
+  StringSet *parent_matches = StringSetNew();
+  if (stack_index > 0 && frame->inherits_previous)
+  {
+    StringSetDestroy(parent_matches);
+    parent_matches = ClassesMatchingLocalRecursive(ctx, regex, tags, first_only, stack_index - 1);
+  }
+  StringSetJoin(matches, parent_matches, xstrdup);
+  StringSetDestroy(parent_matches);
```

Comment:
```suggestion
  if (stack_index > 0 && frame->inherits_previous)
  {
    StringSet *parent_matches = ClassesMatchingLocalRecursive(ctx, regex, tags, first_only, stack_index - 1);
    StringSetJoin(matches, parent_matches, xstrdup);
    StringSetDestroy(parent_matches);
  }

```
The `parent_matches` variable can live within the scope of the if statement

Reply:
ah, right, good one.